### PR TITLE
Backport PR #10048 on branch 5.x

### PR DESCRIPTION
**Pull request by Carreau**
See #10048 / 94d6f5b

### CODE REVIEW

File: IPython/core/display.py [hunk 1]
```diff
@@ -168,7 +211,69 @@ def display(*objs, **kwargs):
...
                 # kwarg-specified metadata gets precedence
                 _merge(md_dict, metadata)
-            publish_display_data(data=format_dict, metadata=md_dict)
+            publish_display_data(data=format_dict, metadata=md_dict, **kwargs)
+    if display_id:
+        return DisplayHandle(display_id)
+
+
+# use * for keyword-only display_id arg
+def update_display(obj, *, display_id, **kwargs):
```
Comment (Carreau) [inline IPython/core/display.py]: Oh, well, that's kwarg only. Need to fix this.

File: IPython/core/displaypub.py [hunk 2]
```diff
@@ -53,7 +53,8 @@ def _validate_data(self, data, metadata=None):
             if not isinstance(metadata, dict):
                 raise TypeError('metadata must be a dict, got: %r' % data)
 
-    def publish(self, data, metadata=None, source=None):
+    # use * to indicate transient, update are keyword-only
+    def publish(self, data, metadata=None, source=None, *, transient=None, update=False, **kwargs):
```
Comment (Carreau) [inline IPython/core/displaypub.py]: kwarg only also.

**Comment by rgbkrk** [comment]
You've made me so happy.

File: IPython/core/display.py [hunk 3]
```diff
@@ -79,7 +80,8 @@ def _display_mimetype(mimetype, objs, raw=False, metadata=None):
 # Main functions
 #-----------------------------------------------------------------------------
 
-def publish_display_data(data, metadata=None, source=None):
+# use * to indicate transient is keyword-only
+def publish_display_data(data, metadata=None, source=None, *, transient=None, **kwargs):
```
Comment (rgbkrk) [inline IPython/core/display.py]: This is the one Python3 feature that was used exclusively here.

**Comment by Carreau** [comment]
Can you have a look at the last commit ?

File: IPython/core/display.py [hunk 4]
```diff
@@ -119,15 +118,18 @@ def publish_display_data(data, metadata=None, source=None, *, transient=None, **
     transient : dict, keyword-only
         A dictionary of transient data, such as display_id.
         """
+
+    sentinel = object()
+    transient = kwargs.get('transient', sentinel)
+    if transient is sentinel:
+        raise TypeError("publish_display_data() missing 1 required keyword-only argument: 'transient'")
```
Comment (rgbkrk) [inline IPython/core/display.py]: For backwards compatibility, do we need to set up an empty transient dict and pass it on instead for `display_data`?

**Comment by Carreau** [comment]
Gosh that was more painful than I thought to backport. 
Test are passing. I'm going to try to re-fall asleep.

**Review by takluyver — APPROVED** [review]

**Comment by rgbkrk** [comment]
Thank you so much for doing this, it will make it easier to adopt inside libraries now.

**Pull request merged**